### PR TITLE
dynamic_rnn time_major=False transpose fixed for inputs of rank>3

### PR DESCRIPTION
**Pull request by ankush-me**
Fix for [dynamic_rnn](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn.py#L682) for inputs of rank > 3 when time_major=False, as discussed in issue #3882.

### CODE REVIEW

**Comment by ankush-me** [comment]
CLA signed.

File: tensorflow/python/ops/rnn.py [hunk 1]
```diff
@@ -785,7 +785,7 @@ def dynamic_rnn(cell, inputs, sequence_length=None, initial_state=None,
 
   if not time_major:
     # (B,T,D) => (T,B,D)
-    flat_input = tuple(array_ops.transpose(input_, [1, 0, 2])
+    flat_input = tuple(array_ops.transpose(input_, [1, 0]+range(2,input_.get_shape().ndims))
```
Comment (ebrevdo) [inline tensorflow/python/ops/rnn.py]: sadly this won't work if input_ has unknown rank.  you'll have to do something like:

input_rank = input_.get_shape().ndims
if input_rank is None:
  transpose_indices = array_ops.concat(0, ([1, 0], array_ops.range(2, array_ops.rank(input_)))
else:
  transpose_indices = [1, 0] + range(2, input_rank)

Comment (ankush-me) [inline tensorflow/python/ops/rnn.py]: Thanks. Got it.

**Comment by ebrevdo** [comment]
(please add unit tests with both known and unknown input rank)

**Comment by ankush-me** [comment]
@ebrevdo -- yes, will do.

**Comment by ebrevdo** [comment]
Any update on this?

**Comment by ankush-me** [comment]
@ebrevdo : Since dynamic_rnn calls _dynamic_rnn_loop which requires that all inputs have [well defined shapes at the graph construction time](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn.py#L917), it is not clear if dynamic_rnn should handle this case where the rank is unknown. 

I patched it as suggested, but then there is no good way of unit-testing it for inputs with unknown shapes, due to the above requirement. What do you suggest?

**Comment by ebrevdo** [comment]
You can probably remove that check. The rnncells themselves can check for
static shape info if needed.

**Comment by ankush-me** [comment]
I removed the shape checks (a2ff968) and added some unit tests (68c874a) -- both with known and unknown ranks of the inputs and outputs at the graph construction time and it seems to work. However, again this required taking out a lot of the shape-checking -- so not sure if we want to keep it as it is, or bring the checks back in. The unit tests would error out if the shape checks are included, hence the tests were not added to to the rnn_test.py file.

**Comment by ebrevdo** [comment]
The only shape checks you need to remove are the lines:

 if not shape[2:].is_fully_defined():
      raise ValueError(
          "Input size (depth of inputs) must be accessible via shape
inference,"
          " but saw value None.")

and these lines (redundant and buggy?):

for input_ in input_shape:
      if input_[1].get_shape() != batch_size.get_shape():
        raise ValueError("All inputs should have the same batch size")

*everything else can stay*

**Comment by ankush-me** [comment]
Correct regarding [this](https://github.com/tensorflow/tensorflow/blob/a2ff968643512b50ea074edcd0561337dad427e2/tensorflow/python/ops/rnn.py#L830); should not have been commented out.

But everything else, due to [line 923](https://github.com/tensorflow/tensorflow/blob/a2ff968643512b50ea074edcd0561337dad427e2/tensorflow/python/ops/rnn.py#L923):
```
const_time_steps, const_batch_size = inputs_got_shape[0].as_list()[:2]
```
where state shape is required, needs to be taken out (?), i.e.: 

1. Lines [930](https://github.com/tensorflow/tensorflow/blob/a2ff968643512b50ea074edcd0561337dad427e2/tensorflow/python/ops/rnn.py#L930) - [938](https://github.com/tensorflow/tensorflow/blob/a2ff968643512b50ea074edcd0561337dad427e2/tensorflow/python/ops/rnn.py#L938):
	```
    got_time_steps = shape[0].value
    got_batch_size = shape[1].value
    if const_time_steps != got_time_steps:
      raise ValueError(
          "Time steps is not the same for all the elements in the input in a "
          "batch.")
    if const_batch_size != got_batch_size:
      raise ValueError(
          "Batch_size is not the same for all the elements in the input.")
	```

2. And, lines [1026](https://github.com/tensorflow/tensorflow/blob/a2ff968643512b50ea074edcd0561337dad427e2/tensorflow/python/ops/rnn.py#L1026) - [1030](https://github.com/tensorflow/tensorflow/blob/a2ff968643512b50ea074edcd0561337dad427e2/tensorflow/python/ops/rnn.py#L1030)
	```
	# Restore some shape information
	for output, output_size in zip(final_outputs, flat_output_size):
		shape = _state_size_with_prefix(output_size, prefix=[const_time_steps, const_batch_size])
		output.set_shape(shape)
    ```

**Comment by martinwicke** [comment]
@ebrevdo, @ankush-me whose turn?

**Comment by ankush-me** [comment]
@martinwicke: @ebrevdo, I guess.

**Comment by martinwicke** [comment]
He's on vacation, many people are. I suppose this will have to wait a little longer.

**Comment by drpngx** [comment]
Happy New Year everyone!

@ankush-me could you sync and push again?

**Comment by ankush-me** [comment]
@drpngx Sorry for the delay. I've synced with master.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
This fails tests. Could you fix?

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Comment by vrv** [comment]
@ankush-me will you have time to look at this again any time soon?  I'll close by the end of the week if not.  Thanks again for your contribution.

**Comment by ankush-me** [comment]
@vrv: there are some assumptions about shapes being fully defined at graph construction time in ` _dynamic_rnn_loop`, as highlighted in my comment posed on Nov 15, 2016 above. Due to these assumptions, I cannot write unit tests with inputs with unknown shapes at graph construction time, as is desired by @ebrevdo. Can I take out these checks (as pointed out in the comment above)? Please let me know. Thanks.

**Comment by ebrevdo** [comment]
Can you modify the checks to only run it the static values are available?

**Comment by vrv** [comment]
@ankush-me any updates?  I'm going to close by the end of the day to keep the PR list small, but feel free to send a new PR if you get around to fixing this.

**Comment by ankush-me** [comment]
@vrv I should find time over the weekend to fix this. If you can wait till Monday, that'll be great.

**Comment by ankush-me** [comment]
Done: removed the unnecessary shape checks and added unit tests for input tensors of rank > 3 in kernel_tests/rnn_test.py for both known and unknown static rank. Can you please test this?

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
Jenkins, test this please.

Are you done pushing?

**Comment by ankush-me** [comment]
Yes.

**Comment by ankush-me** [comment]
Fixed bugs. Tests seem to pass locally. Can you please test this again?

File: tensorflow/python/kernel_tests/rnn_test.py [hunk 2]
```diff
@@ -33,6 +33,8 @@
 from tensorflow.core.protobuf import config_pb2
 from tensorflow.python.client import session
 from tensorflow.python.framework import dtypes
+from tensorflow.python.framework import tensor_shape
```
Comment (ebrevdo) [inline tensorflow/python/kernel_tests/rnn_test.py]: imports should be alphabetically ordered.

File: tensorflow/python/kernel_tests/rnn_test.py [hunk 3]
```diff
@@ -63,6 +65,40 @@ def __call__(self, input_, state, scope=None):
     return (input_ + 1, state + 1)
 
 
+class DummyMultiDimensionalLSTM(rnn_cell_impl._RNNCell):
```
Comment (ebrevdo) [inline tensorflow/python/kernel_tests/rnn_test.py]: append to line: "  # pylint: disable=protected-access"

File: tensorflow/python/kernel_tests/rnn_test.py [hunk 4]
```diff
@@ -63,6 +65,40 @@ def __call__(self, input_, state, scope=None):
...
+  The input to this cell may have an arbitrary number of dimensions that follow
+  the preceding 'Time' and 'Batch' dimensions.
+  """
+
+  def __init__(self, dims):
+    """Initialize the Multi-dimensional LSTM cell.
+
+    Args:
+      dims: tuple that contains the dimensions of the output of the cell,
+      without including 'Time' or 'Batch' dimensions.
```
Comment (ebrevdo) [inline tensorflow/python/kernel_tests/rnn_test.py]: Raises: section.

File: tensorflow/python/kernel_tests/rnn_test.py [hunk 5]
```diff
@@ -554,6 +590,52 @@ def benchmarkStaticUnrollStateConcatVsStateTuple(self):
                 iters=20,
                 wall_time=t_dt)
 
+class DynamicRNNInputND(test.TestCase):
```
Comment (ebrevdo) [inline tensorflow/python/kernel_tests/rnn_test.py]: this is a test, the class name should end with "Test"

File: tensorflow/python/ops/rnn.py [hunk 6]
```diff
@@ -23,6 +23,7 @@
 from tensorflow.python.framework import dtypes
 from tensorflow.python.framework import ops
 from tensorflow.python.framework import tensor_shape
+
```
Comment (ebrevdo) [inline tensorflow/python/ops/rnn.py]: remove unnecessary whitespace

File: tensorflow/python/ops/rnn.py [hunk 7]
```diff
@@ -485,8 +486,18 @@ def dynamic_rnn(cell, inputs, sequence_length=None, initial_state=None,
...
   if not time_major:
     # (B,T,D) => (T,B,D)
-    flat_input = tuple(array_ops.transpose(input_, [1, 0, 2])
-                       for input_ in flat_input)
+    def _get_transpose_indices(x):
+      x = ops.convert_to_tensor(x)
+      x_rank = x.get_shape().ndims
+      if x_rank is None:
+        x_rank = array_ops.rank(x)
+        indices = array_ops.concat([[1,0],math_ops.range(2,x_rank)],0)
```
Comment (ebrevdo) [inline tensorflow/python/ops/rnn.py]: spaces after every comma, here and elsewhere.

File: tensorflow/python/ops/rnn.py [hunk 8]
```diff
@@ -485,8 +486,18 @@ def dynamic_rnn(cell, inputs, sequence_length=None, initial_state=None,
...
+      else:
+        indices = [1,0] + range(2,x_rank)
+      return indices
+
+    flat_input = tuple(array_ops.transpose(input_,
```
Comment (ebrevdo) [inline tensorflow/python/ops/rnn.py]: newline / indentation are weird; please fix while keeping w/in 80 char limit.

File: tensorflow/python/ops/rnn.py [hunk 9]
```diff
@@ -550,8 +561,8 @@ def _assert_has_shape(x, shape):
     if not time_major:
       # (T,B,D) => (B,T,D)
       flat_output = nest.flatten(outputs)
-      flat_output = [array_ops.transpose(output, [1, 0, 2])
-                     for output in flat_output]
+      flat_output = [array_ops.transpose(output,
+                     _get_transpose_indices(output)) for output in flat_output]
```
Comment (ebrevdo) [inline tensorflow/python/ops/rnn.py]: ditto

File: tensorflow/python/ops/rnn.py [hunk 10]
```diff
@@ -611,22 +622,7 @@ def _dynamic_rnn_loop(cell,
...
-  const_time_steps, const_batch_size = inputs_got_shape[0].as_list()[:2]
-
-  for shape in inputs_got_shape:
-    if not shape[2:].is_fully_defined():
-      raise ValueError(
-          "Input size (depth of inputs) must be accessible via shape inference,"
-          " but saw value None.")
-    got_time_steps = shape[0].value
-    got_batch_size = shape[1].value
-    if const_time_steps != got_time_steps:
```
Comment (ebrevdo) [inline tensorflow/python/ops/rnn.py]: removing these things is weird; we're just checking consistency across multiple input tensors.  why do you need to remove these lines?  perhaps these ifs should only happen if const_time_steps and got_time_steps are both not None?

File: tensorflow/python/ops/rnn.py [hunk 11]
```diff
@@ -611,22 +622,7 @@ def _dynamic_rnn_loop(cell,
   inputs_got_shape = tuple(input_.get_shape().with_rank_at_least(3)
                            for input_ in flat_input)
 
-  const_time_steps, const_batch_size = inputs_got_shape[0].as_list()[:2]
-
-  for shape in inputs_got_shape:
-    if not shape[2:].is_fully_defined():
```
Comment (ebrevdo) [inline tensorflow/python/ops/rnn.py]: i suppose this should be: "if shape.ndims is None or shape.ndims < 2:"

**Review by ebrevdo — CHANGES_REQUESTED** [review]

**Comment by ankush-me** [comment]
@ebrevdo Thanks for the review! Will make these changes soon.

File: tensorflow/python/ops/rnn.py (hunk 8 above)
Comment (ankush-me) [inline tensorflow/python/ops/rnn.py]: Not exactly sure if I understand this correctly. Can you please check again?

File: tensorflow/python/ops/rnn.py (hunk 11 above)
Comment (ankush-me) [inline tensorflow/python/ops/rnn.py]: Since we want this to run even when the static rank is unknown, these checks should be removed, right? As also discussed [here](https://github.com/tensorflow/tensorflow/pull/5142#issuecomment-260748149).

File: tensorflow/python/ops/rnn.py (hunk 10 above)
Comment (ankush-me) [inline tensorflow/python/ops/rnn.py]: I've put these back in. 
`const_time_steps` and `const_batch_size` are taken from the first input; not clear if that is a good idea, if the first input has these values as `None`.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by ebrevdo** [comment]
We actually have an internal piece of code we're gonna move over here to
fix this problem; if that's OK:

https://github.com/tensorflow/tensorflow/blob/34df7ee6f5a8f931d2433dc7e6e739bc880a35ea/tensorflow/contrib/seq2seq/python/ops/decoder.py#L41

So perhaps we can hold off on this PR until that code is moved internally?
Because then it can be used here and in the dynamic decoder.

**Comment by ankush-me** [comment]
Ok.

**Comment by marcociccone** [comment]
any update on this PR? When do you think it's going to be merged? 
for reference: @carlthome

**Comment by ebrevdo** [comment]
I just pushed a fix internally; it should show up on master
today/tomorrow.  It's safe to close this PR.

**Comment by carlthome** [comment]
Wonderful! Hope this makes it onto PyPI ASAP! Thanks for notifying @marcociccone.
